modules/userdirectory-canvas/src/main/java/org/opencastproject/userdirectory/canvas/CanvasUserProviderInstance.java

```diff
@@ -0,0 +1,705 @@
+          return false;
+        }
+      } catch (PatternSyntaxException e) {
+        logger.warn("Invalid regular expression for course pattern {} - disabling checks, exception: {}", coursePattern, e);
+        coursePattern = null;
+      }
+
+      int code;
+      String urlString = canvasUrl + "/api/v1/courses/" + courseId;
+      try {
+          URL url = new URL(urlString);
+          logger.debug("Verifying course: {} using API: {}", courseId, url.toString());
+          HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+          connection.setRequestMethod("GET");
+          connection.setRequestProperty("Authorization", "Bearer " + canvasToken);
+          connection.setRequestProperty("User-Agent", OC_USERAGENT);
+
+          connection.connect();
+          code = connection.getResponseCode();
+      } catch (Exception e) {
+          logger.warn("Exception verifying Canvas course {} at {}: {}", courseId, urlString, e);
+          return false;
+      }
+      // HTTP OK 200 for course exists, return false for everything else (typically 404 not found)
+      return (code == 200);
+  }
+
+  /**
+   * Get user roles from Canvas
+   * Query with /api/v1/users/sis_user_id:{userId}/courses and traverse
+   * pagination
+   *
+   * @param courseId
+   */
+  private String[] getRolesFromCanvas(String userId) {
+    logger.debug("getRolesFromCanvas({})", userId);
+    String nextLinkRegex = "(?:.*)<(.+)>; rel=\"next\"";
+    Pattern pattern = Pattern.compile(nextLinkRegex);
+
+    List<String> roleList = new ArrayList<String>();
+
+    /*
+    Official Canvas documentation on API pagination states that the maximum result set per page
+    is 10 by default. A 'per_page' querystring can be added to yield more results, however the
+    maximums aren't documentated, Unofficially it appears that the courses API returns a maximum
+    of 100 results. Link response headers for navigation are returned when more results are available.
+    */
+    String nextPage = canvasUrl + "/api/v1/users/sis_user_id:" + userId + "/courses?per_page=100";
+    try {
+      while (StringUtils.isNotBlank(nextPage)) {
+        URL url = new URL(nextPage);
+        logger.debug("Requesting courses for user:{}, using API: {}", userId, url.toString());
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestMethod("GET");
+        connection.setDoOutput(true);
+        connection.setRequestProperty("Authorization", "Bearer " + canvasToken);
+        connection.setRequestProperty("User-Agent", OC_USERAGENT);
+
+        String json = IOUtils.toString(new BufferedInputStream(connection.getInputStream()));
+        logger.debug(json);
+
+        logger.debug("Response code for API: {}, is: {}", url.toString(), connection.getResponseCode());
+        JsonArray courses = new JsonParser().parse(json).getAsJsonArray();
```

Comment:
Nothing wrong with it, I'm just wondering if there was a specific reason to go with GSON's [JsonParser](http://tutorials.jenkov.com/java-json/gson-jsonparser.html) instead of the [Gson JSON parser](http://tutorials.jenkov.com/java-json/gson.html).

Also, I know that Gson is thread-safe and it makes a lot of sense to initialize the parser once as a static class variable since the initialization is pretty expensive. I would expect the same for the JsonParser. But then, I've never used it and ;-)